src/mlpack/methods/isomap/isomap.hpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file isomap.hpp
+ * @author Rishabh Ranjan
+ * 
+ * This file defines the Isomap class which implements ISOMAP algorithm on a
+ * given dataset. The templates are provided so that this class can include
+ * variations of constructing neighbourhood graphs and finding all pair
+ * shortest paths in the graph. Presently, neighbourhood graph is constructed
+ * using only K_Nearest_Neighbours and all pair shortest path is calcuated
+ * using only Dijkstra's Algorithm
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ISOMAP_ISOMAP_HPP
+#define MLPACK_METHODS_ISOMAP_ISOMAP_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/metrics/lmetric.hpp>
+#include "k_nearest.hpp"
+#include "dijkstra.hpp"
+#include "mds.hpp"
+
+namespace mlpack {
+namespace isomap {
+
+/**
+ * This class implements the Isomap algorithm which is part of Manifold
+ * Learning algorithms, and is used for dimensionality reduction. This
+ * algorithm converts the dataset into a lower dimensional space which
+ * it is assumed to belong from the higher dimensional space it is
+ * currently embedded into.
+ * This implementation is based on the paper:
+ * Lawrence Cayton, 2005, 17, Algorithms for Manifold Learning
+ */
+
+template <typename NeighbourhoodFunction = K_Nearest,
+          typename ShortestPathAlgo = Dijkstra>
+class Isomap
+{
+ public:
+  /**
+   * Create the Isomap obejct and set the n_neighbours parameter to
+   * specify the number of neighbours to consider while constructing
+   * neighbourhood graph.
+   * 
+   * @param n_neighbours -number of neighbours to consider
+   *                      for neighbourhood graph
+   */
+
+  Isomap(const size_t n_neighbours,
+         const NeighbourhoodFunction& neighbourhood = NeighbourhoodFunction(),
+         const ShortestPathAlgo& shortestPath = ShortestPathAlgo()) :
+         n_neighbours(n_neighbours),
+         neighbourhood(neighbourhood),
+         shortestPath(shortestPath)
+  { }
+
+  /**
+   * This is the main driver function to perform Isomap. Just the input
+   * matrix is required and it is safe to pass it as reference.
+   * 
+   * @param input -the input dataset to perform Isomap on.
+   */
+  void Apply(arma::mat& input)
+  {
+    arma::mat disMat(input.n_cols, input.n_cols);
+
+    // calculating distance matrix from the given input matrix
+    CalcDistanceMatrix(input, disMat);
+
+    // constructing neighbourhood graph (K_nearest is used)
+    neighbourhood.MakeNeighbourhoodGraph(disMat, n_neighbours);
+
+    // finding all pair shortest path in the neighbourhood graph created
+    shortestPath.FindShortestPath(disMat);
```

Comment:
The goal of the entire algorithm up to this point is to construct a matrix that contains the shortest geodesic path between each pair of points.  So Dijkstra's algorithm is used here, but what you could actually equivalently use is just a minimum spanning tree.  We already have code that can compute that (as part of the nearest neighbor search) in the `DualTreeBoruvka` class.  So I would suggest that perhaps we should use that instead of the hand-implemented Dijkstra's algorithm here.

Reply:
As far as I know MST is a good approximation for all pair shortest paths in cases of sparse graphs as then it includes most of the useful edges. But I am not sure how accurate it would be in case of dense graphs (the type of graph would depend on the number of neighbours). If you think the approximation is reasonable we can use MST for this. Also, the complexity would reduce a little with MST.

Reply:
Also, K-Nearest would return a directed graph. For applying Boruvka's algorithm we would have to change it to undirected graph, which could increase the number of nearest neighbours present in the graph for a node. But I think that would not affect the working of Isomap as anyway we make the geodesic distance matrix symmetric before cMDS.